### ISSUE TITLE
Ledger meta-data URL validity

### ISSUE THREAD

**Issue by vantuz-subhuman** [input-output-hk/fm-ledger-rules#289]
Hello! I have a little question about the section of the delegation design spec that describes how wallets should treat the meta-data URL for a pool:

https://github.com/input-output-hk/fm-ledger-rules/blob/409e46a4ede339975bd5d64dc5498e93bd62a3ce/docs/delegation_design_spec/delegation_design_spec.tex#L1034-L1046

![image](https://user-images.githubusercontent.com/5585355/53878441-f084ae80-401c-11e9-9e1d-81765a6450c3.png)

The question: is there any specific requirements/recommendations on how wallets should treat caching for this data? E.g. when a wallet encounters a new pool certificate - it may download the data from the URL, validate it successfully, and store locally. At a later date when user goes thru the list of pools - wallet will display it, as there's a valid meta-data available for it, but it does not guarantee that the URL still returns the same valid data.

Would it be considered a bad case, and should the wallet make sure that URL itself is working to display a pool? Or would it be an ok case to have pool listed just because this particular wallet has the valid data available (local view)? In case of a resync the same wallet would not display this pool anymore, of course.

Thank you!

@kantp

**Comment by vantuz-subhuman** [comment]
Question number 2 is the size of this meta-data. Section describing the meta-data schema says "A short textual description":

https://github.com/input-output-hk/fm-ledger-rules/blob/409e46a4ede339975bd5d64dc5498e93bd62a3ce/docs/delegation_design_spec/delegation_design_spec.tex#L2262

![image](https://user-images.githubusercontent.com/5585355/53878915-2d04da00-401e-11e9-994c-fce4e3dff076.png)

Wallets would definitely want to have some sanity size-limit defined for this meta-data so that when they download it from an **UNTRUSTED** url - they don't have to stream a gigabyte of description, and can fail-fast if it exceeds the limit even by 1 byte.

Thank you!

**Comment by kantp** [comment]
Hi Vantuz, thanks for your questions!

Regarding your second question, we decided to just place an overall size limit on the metadata (512 bytes), instead of on individual fields.

About the first question: note that we do not expect individual wallets to query all those URLs, but have stake pool metadata servers in the design, that do the work of caching the metadata of different pools, and then allowing wallets to do bulk lookup instead of querying a large number of (potentially slow) URLs (of course, if a wallet prefers to get the data from the pools directly, it is free to do so).
Still, the question of cache invalidation remains, of course. I'd say displaying a pool where the metadata URL is temporarily unavailable, or has recently gone offline, is acceptable. But we should periodically refresh the cache (whether it's in the metadata server or in the wallet) to make sure that pools that no longer serve their metadata are pruned; it's reasonable to expect that if a pool stopped serving their metadata, they are also likely to have stopped running their node.

**Comment by kantp** [comment]
Does that answer your questions, @vantuz-subhuman?

**Comment by vantuz-subhuman** [comment]
Thank you, Philipp! Yes, it answers it quite well and I have missed the part about the 512 byte limit, so this is good! 😄 

Am I getting it right that meta-data servers (as planned) will download data from all external URLs and then export a special API that allows anyone to query data by pool IDs, and maybe also in butches? Are they planned to provide the meta-data hosting itself, so that pool creator can upload their meta-data to one of these servers, instead of looking for a 3rd party external file hosting? 🤔

**Comment by vantuz-subhuman** [comment]
@kantp, I also wonder if no limit on individual meta-data field means that a pool with a nearly 500 bytes long ticker is a valid case? )

**Comment by vantuz-subhuman** [comment]
About the meta-data servers. This section makes it a little confusing:

![image](https://user-images.githubusercontent.com/5585355/53881842-bcfa5200-4025-11e9-8275-5d469f74a2a0.png)

Does this mean that 3rd parties will be encouraged to **develop** their own proxy servers and share code and binaries?

**Comment by kantp** [comment]
The current plan is to have them working only as a cache (and a possibility to filter out pools where the URL to the homepage is malicious (phishing, ...). Pool operators will already run a node, and most likely also a website, so one additional endpoint for the metadata shouldn't be that much of a problem, I guess.

**Comment by kantp** [comment]
> @kantp, I also wonder if no size-limit on individual meta-data fields means that a pool with a nearly 500 bytes long ticker is a valid case? )

Oh no, the ticker should be restricted to 3-5 chars. This is missing in the document. Good catch!

**Comment by kantp** [comment]
> Does this mean that 3rd parties will be encouraged to **develop** their own proxy servers and share code and binaries?

No, the plan is that we will provide the code for the proxy servers, and anyone can then run one.

**Comment by vantuz-subhuman** [comment]
Thank you, Philipp, just wanted to clarify about the ticker length in relation to the comment about no limit on individual fields. And also clarify that I read the section about servers right 🙂 🙌
